keras_nlp/layers/token_and_position_embedding.py

```diff
@@ -0,0 +1,106 @@
+
+    This class assumes that in the input tensor, the last dimension corresponds
+    to the features, and the dimension before the last corresponds to the
+    sequence.
+
+    Args:
+        `vocabulary_size`: The size of the vocabulary (should be no larger than 999)
+        `max_length`: The maximum length of input sequence
+        `embedding_dim`: The output dimension of the embedding layer
+        `embeddings_initializer`: The initializer to use for the Embedding Layer
+        `embeddings_regularizer`: The regularizer to user for regularization of the Embedding Layer
+        `mask_zero`: Boolean, whether or not the input value 0 is a special "padding" value that should be masked out.
+            This is useful when using recurrent layers which may take variable length input. If this is True, then all
+            subsequent layers in the model need to support masking or an exception will be raised. If mask_zero is set
+            to True, as a consequence, index 0 cannot be used in the vocabulary
+            (input_dim should equal size of vocabulary + 1).
+
+    Example:
+    ```python
+    token_and_position_embed = keras_nlp.layers.TokenAndPositionEmbedding(
+        vocabulary_size=vocab_size,
+        max_length=max_length,
+        embedding_dim=embed_dim,
+    )
+    outputs = token_and_position_embed(inputs)
+    ```
+    """
+
+    def __init__(
+        self,
+        vocabulary_size,
+        max_length,
+        embedding_dim,
+        embeddings_initializer="glorot_uniform",
+        embeddings_regularizer=None,
+        mask_zero=False,
+        **kwargs
+    ):
+        super().__init__()
+        if vocabulary_size is None:
+            raise ValueError(
+                "`vocabulary_size` must be an Integer, not `None`."
+            )
+        if max_length is None:
+            raise ValueError("`max_length` must be an Integer, not `None`.")
+        if embedding_dim is None:
+            raise ValueError("`embedding_dim` must be an Integer, not `None`.")
+        self.vocabulary_size = int(vocabulary_size)
+        self.max_length = int(max_length)
+        self.embedding_dim = int(embedding_dim)
+        self.token_embeddings = keras.layers.Embedding(
+            vocabulary_size,
+            embedding_dim,
+            embeddings_initializer=embeddings_initializer,
+            embeddings_regularizer=embeddings_regularizer,
+            mask_zero=mask_zero,
+        )
+        self.position_embeddings = keras_nlp.layers.PositionEmbedding(
+            max_length=max_length,
+        )
+
+    def get_config(self):
+        config = super().get_config()
+        config.update(
+            {
+                "vocabulary_size": self.vocabulary_size,
+                "max_length": self.max_length,
+                "embedding_dim": self.embedding_dim,
+            },
+            **self.token_embeddings.get_config()
+        )
+        del config["name"]
+        return config
+
+    def build(self, input_shape):

```

Comment:
We shouldn't need this if all we do is chain to super

Reply:
If we want to do something like this:
```python
config = super().get_config()
config.update(
    # {
    #     "vocabulary_size": self.vocabulary_size,
    #     "max_length": self.max_length,
    #     "embedding_dim": self.embedding_dim,
    # },
    **self.token_embeddings.get_config(),
    **self.position_embeddings.get_config()
)
```
instead of updating `config` with class variables or with individual objects `get_config()` method the test fails for `name` attribute. Both layers have their own `name` attribute, and test will fail when we want to update `name` key with two names.
Otherwise if using only `position_embeddings` `get_config()` method, and using class attributes (commented section) the method again fails the test as it sets the `name` attribute to `embedding_1` while test assumes it to be `embedding`.

What is the correct way? Should we give name ourselves?

Reply:
We should inherit name from super().get_config(), and not do any `**` dumping of the child layer configs into our config. Basically, our config should agree exactly with the init args we handle, and can pull from child layers if we don't store the param on our layer. Incomplete, but something like this...

```
config = super().get_config()
config.update(
    {
        "vocabulary_size": self.vocabulary_size,
        "max_length": self.max_length,
        "embedding_dim": self.embedding_dim,
        "mask_zero": self.token_embeddings.mask_zero,
        "embeddings_initializer": keras.initializers.serialize(self.tokens_embeddings.embeddings_initializer),
    },
)
```

Does that work?

Reply:
Yeah that will do!